### PR TITLE
Work.v00004

### PR DESCRIPTION
**Pull request by Artificial-Friend**

### CODE REVIEW

**Review by Vedroid — APPROVED** [review]

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    @Override
+    public boolean delete(Long id) {
+        return dao.delete(id);
+    }
+
+    @Override
+    public void addDriverToCar(Driver driver, Car car) {
+        List<Driver> drivers = car.getDrivers();
+        drivers.add(driver);
+        car.setDrivers(drivers);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]:
```suggestion
        car.getDrivers().add(driver);
```

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        dao.update(car);
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        if (car.getDrivers().contains(driver)) {
+            List<Driver> drivers = car.getDrivers();
+            drivers.remove(driver);
+            car.setDrivers(drivers);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]:
```suggestion
            car.getDrivers().remove(driver);
```

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        return getAll()
+                .stream()
+                .filter(car -> car.getDrivers()
+                        .stream()
+                        .map(Driver::getId)
+                        .collect(Collectors.toList()).contains(driverId))
+                .collect(Collectors.toList());
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]: move this logic to the dao layer, when you connect db to the project you will receive sorted data from db already

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public void addDriverToCar(Driver driver, Car car) {
+        List<Driver> drivers = car.getDrivers();
+        drivers.add(driver);
+        car.setDrivers(drivers);
+        dao.update(car);
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        if (car.getDrivers().contains(driver)) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]: real db will handle this check and throw exception if no driver is present, so let's remove it

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        daoCar.update(car);
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        Optional<Driver> driver = daoDriver.get(driverId);
+        List<Car> cars = daoCar.getAll();
+        cars.removeIf(car -> !car.getDrivers().contains(driver));
+        return cars;
+    }
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]: as i've mentioned logic of filtering cars should be done on the dao layer ( this method is similar to getById) 
here you should just call carDao.getAllByDriver()

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java (hunk 5 above)
Comment (Artificial-Friend) [reply src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]: OK, sorry, misinterpreted the task.

**Review by RomanDubovskyi — APPROVED** [review]

File: src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java (hunk 5 above)
Comment (RomanDubovskyi) [reply src/main/java/core/basesyntax/service/implementations/CarServiceImpl.java]: all good)

**Pull request merged**